### PR TITLE
git-credential-manager: init at 2.1.2

### PR DESCRIPTION
**Pull request by 999eagle**
###### Description of changes

Add [git-credential-manager](https://github.com/git-ecosystem/git-credential-manager) package, a secure, cross-platform Git credential storage with authentication to GitHub, Azure Repos, and other popular Git hosting services. 

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+buildDotnetModule rec {
+  pname = "git-credential-manager";
+  version = "2.1.2";
+
+  src = fetchFromGitHub {
+    owner = "git-ecosystem";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-PeQ9atSCgSvduAcqY2CnNJH3ucvoInduA5i8dPUJiHo=";
+  };
```
Comment (AndersonTorres) [inline pkgs/applications/version-management/git-credential-manager/default.nix:27]:
```suggestion
  };

```

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+, libSM
+, fontconfig
+, libsecret
+, gnupg
+, pass
+
+, withGuiSupport ? true
+, withLibsecretSupport ? true
+, withGpgSupport ? true
+}:
```
Comment (AndersonTorres) [inline pkgs/applications/version-management/git-credential-manager/default.nix:16]:
```suggestion
}:

```

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+, buildDotnetModule
+, dotnetCorePackages
+, libX11
+, libICE
+, libSM
+, fontconfig
+, libsecret
+, gnupg
+, pass
+
```
Comment (AndersonTorres) [inline pkgs/applications/version-management/git-credential-manager/default.nix:12]:
```suggestion
```

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+assert withLibsecretSupport -> withGuiSupport;
+buildDotnetModule rec {
+  pname = "git-credential-manager";
+  version = "2.1.2";
+
+  src = fetchFromGitHub {
+    owner = "git-ecosystem";
+    repo = pname;
+    rev = "v${version}";
+    sha256 = "sha256-PeQ9atSCgSvduAcqY2CnNJH3ucvoInduA5i8dPUJiHo=";
```
Comment (AndersonTorres) [inline pkgs/applications/version-management/git-credential-manager/default.nix:26]:
```suggestion
    hash = "sha256-PeQ9atSCgSvduAcqY2CnNJH3ucvoInduA5i8dPUJiHo=";
```

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+, withGpgSupport ? true
+}:
+assert withLibsecretSupport -> withGuiSupport;
+buildDotnetModule rec {
+  pname = "git-credential-manager";
+  version = "2.1.2";
+
+  src = fetchFromGitHub {
+    owner = "git-ecosystem";
+    repo = pname;
```
Comment (AndersonTorres) [inline pkgs/applications/version-management/git-credential-manager/default.nix:24]:
```suggestion
    repo = "git-credential-manager";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by wrmilling** [comment]
Tested building on NixOS x86_64-linux, successfully builds. Diagnostic command runs successfully and was able to test the GUI popup for a credential via `git send-email`. LGTM. 

```
/nix/store/ryskig8s0zgipr1l5lp36mg2kicf4mwh-git-credential-manager-2.1.2/bin/git-credential-manager diagnose
Running diagnostics...

 [ OK ] Environment
 [ OK ] File system
 [ OK ] Networking
 [ OK ] Git
 [ OK ] Credential storage
 [ OK ] Microsoft authentication (AAD/MSA)
 [ OK ] GitHub API

Diagnostic summary: 7 passed, 0 skipped, 0 failed.
```

**Review by wrmilling — APPROVED** [review]

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  };
+
+  projectFile = "src/shared/Git-Credential-Manager/Git-Credential-Manager.csproj";
+  nugetDeps = ./deps.nix;
+  dotnet-sdk = dotnetCorePackages.sdk_6_0;
+  dotnet-runtime = dotnetCorePackages.runtime_6_0;
+  dotnetInstallFlags = [ "--framework" "net6.0" ];
+  executables = [ "git-credential-manager" ];
+
+  runtimeDeps = [ fontconfig ] ++ (lib.optionals withGuiSupport [ libX11 libICE libSM ]) ++ (lib.optional withLibsecretSupport libsecret);
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/git-credential-manager/default.nix:36]:
```suggestion
  runtimeDeps = [ fontconfig ]
    ++ lib.optionals withGuiSupport [ libX11 libICE libSM ]
    ++ lib.optional withLibsecretSupport libsecret;
```

File: pkgs/applications/version-management/git-credential-manager/default.nix [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  makeWrapperArgs = lib.optionals withGpgSupport [ "--prefix" "PATH" ":" (lib.makeBinPath [ gnupg pass ]) ];
+
+  passthru.updateScript = ./update.sh;
+
+  meta = with lib; {
+    description = "Secure, cross-platform Git credential storage with authentication to GitHub, Azure Repos, and other popular Git hosting services.";
+    homepage = "https://github.com/git-ecosystem/git-credential-manager";
+    license = with licenses; [ mit ];
+    platforms = with platforms; [ linux darwin ];
```
Comment (SuperSandro2000) [inline pkgs/applications/version-management/git-credential-manager/default.nix:45]: I think that is the default but I am not fully sure https://github.com/NixOS/nixpkgs/blob/master/pkgs/build-support/dotnet/build-dotnet-module/default.nix#L89-L92

Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-credential-manager/default.nix:45]: No, it isn't

```
       error: Package ‘git-credential-manager-2.1.2’ in /nix/store/17jvfrx5c9clfv5zzl0qgk5i93szzj9i-source/pkgs/applications/version-management/git-credential-manager/default.nix:42 is not available on the requested hostPlatform:
         hostPlatform.config = "x86_64-unknown-linux-gnu"
         package.meta.platforms = [ ]
         package.meta.badPlatforms = [ ]
       , refusing to evaluate.

       a) To temporarily allow packages that are unsupported for this system, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_UNSUPPORTED_SYSTEM=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowUnsupportedSystem = true; }
       in configuration.nix to override this.

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
         { allowUnsupportedSystem = true; }
       to ~/.config/nixpkgs/config.nix.
```

Comment (SuperSandro2000) [reply pkgs/applications/version-management/git-credential-manager/default.nix:47]:
```suggestion
    platforms = platforms.unix;
```
that should do the trick

**Review by mdarocha — APPROVED** [review]

**Comment by lostmsu** [comment]
How do I use this?

**Comment by 999eagle** [comment]
please don't resurrect old prs like this and this isn't really relevant to nix at all.

still, the way to use this is to set `credential.helper` in your git config to the path to the gcm binary. if using home-manager, this would be done for example by setting `programs.git.extraConfig.credential.helper = "${pkgs.git-credential-manager}/bin/git-credential-manger";`.

**Pull request merged**